onnxruntime/core/session/provider_bridge_ort.cc

```diff
@@ -0,0 +1,563 @@
+  void set_s(const ::std::string& value) override { v_.set_s(value); }
+  const ::std::string& s() const override { return v_.s(); }
+  void set_name(const ::std::string& value) override { v_.set_name(value); }
+  void set_type(::onnx::AttributeProto_AttributeType value) override { v_.set_type(value); }
+  ::onnx::Provider_TensorProto* add_tensors() override {
+    // Kind of a hack, but the pointer is only valid until the next add_tensors call
+    tensors_ = onnxruntime::make_unique<Provider_TensorProto_Impl>(v_.add_tensors());
+    return tensors_.get();
+  }
+
+  ONNX_NAMESPACE::AttributeProto v_;
+  std::unique_ptr<Provider_TensorProto_Impl> tensors_;
+};
+
+struct Provider_KernelDef_Impl : Provider_KernelDef {
+  Provider_KernelDef_Impl(std::unique_ptr<KernelDef> p) : p_(std::move(p)) {}
+  std::unique_ptr<KernelDef> p_;
+};
+
+struct Provider_KernelDefBuilder_Impl : Provider_KernelDefBuilder {
+  Provider_KernelDefBuilder& SetName(const char* op_name) override {
+    v_.SetName(op_name);
+    return *this;
+  }
+  Provider_KernelDefBuilder& SetDomain(const char* domain) override {
+    v_.SetDomain(domain);
+    return *this;
+  }
+
+  Provider_KernelDefBuilder& SinceVersion(int since_version) override {
+    v_.SinceVersion(since_version);
+    return *this;
+  }
+  Provider_KernelDefBuilder& Provider(const char* provider_type) override {
+    v_.Provider(provider_type);
+    return *this;
+  }
+  Provider_KernelDefBuilder& TypeConstraint(const char* arg_name, MLDataType supported_type) override {
+    v_.TypeConstraint(arg_name, supported_type);
+    return *this;
+  }
+
+  std::unique_ptr<Provider_KernelDef> Build() override {
+    return onnxruntime::make_unique<Provider_KernelDef_Impl>(v_.Build());
+  }
+
+  KernelDefBuilder v_;
+};
+
+struct Provider_NodeArg_Impl : Provider_NodeArg {
+  Provider_NodeArg_Impl(const NodeArg* p) : p_{p} {
+    if (p_->Shape())
+      tensor_shape_proto_.dim_size_ = p_->Shape()->dim_size();
+  }
+
+  const std::string& Name() const noexcept override { return p_->Name(); }
+  const ONNX_NAMESPACE::Provider_TensorShapeProto* Shape() const override { return &tensor_shape_proto_; }
+  virtual ONNX_NAMESPACE::DataType Type() const noexcept override { return p_->Type(); }
+
+  const NodeArg* p_;
+  ONNX_NAMESPACE::Provider_TensorShapeProto tensor_shape_proto_;
+};
+
+struct Provider_Node_Impl : Provider_Node {
+  Provider_Node_Impl(const Node* p) : p_{p} {}
+  ~Provider_Node_Impl() override {
+    for (auto p : input_defs_)
+      delete p;
+    for (auto p : output_defs_)
+      delete p;
+  }
+
+  const std::string& OpType() const noexcept override { return p_->OpType(); }
+  //  const ONNX_NAMESPACE::OpSchema* Op() const noexcept
+
+  ConstPointerContainer<std::vector<Provider_NodeArg*>> InputDefs() const noexcept override {
+    if (input_defs_.empty()) {
+      for (auto p : p_->InputDefs())
+        input_defs_.push_back(new Provider_NodeArg_Impl(p));
+    }
```

Comment:
If we're going to put the pointers in a vector could this just return gsl::span<const NodeArg*> instead of needing ConstPointerContainer?

Reply:
The original provider interfaces use these, and I want to minimize changes to providers for now (so it's less work to convert them).